<a href="https://colab.research.google.com/github/Collin-Campbell/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_t_tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats 

column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
                'engine-location', 'wheel-base', 'length', 'width', 'height', 
                'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 
                'fuel-system', 'bore', 'stroke', 'compression-ratio', 
                'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df = pd.read_csv('imports-85.data', names = column_headers)

print(df.shape)
df.head()

(205, 26)


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [17]:
df = df.replace({"?": np.NaN})

print(df.shape)
df.head()

(205, 26)


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estimated sample mean to the population mean. Can you say that they are different? 

In [19]:
sample = df.sample(10, random_state=30 )

sample

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
103          0               108         nissan  ...       19          25  13499
26           1               148          dodge  ...       31          38   7609
20           0                81      chevrolet  ...       38          43   6575
155          0                91         toyota  ...       27          32   8778
185          2                94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

In [20]:
sample['highway-mpg'].mean()

32.8

In [21]:
df['highway-mpg'].mean()

30.75121951219512

## Null Hypothesis:

The sample mean for highway-mpg is equal to the population mean (~30.75 mpg) for highway-mpg.

## Alternative Hypothesis:

The sample mean is not equal to the population mean of ~30.75 mpg.

## Confidence Level:  95% 

In [22]:
stats.ttest_1samp(sample['highway-mpg'], 30.75, nan_policy='omit')

Ttest_1sampResult(statistic=0.8917727404948264, pvalue=0.3957307423595724)

## Conclusion:  

Due to a t-statistic of 0.892, and a p-value of 0.396, we fail to reject the null hypothesis that the sample mean for highway-mpg is equal to the population mean (~30.75 mpg) for highway-mpg.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

## Null Hypothesis:

The sample mean for highway-mpg is equal to what the salesman says (35 mpg).

## Alternative Hypothesis:

The sample mean is not equal to what the salesman says (35 mpg).

## Confidence Level:  95% 

In [23]:
stats.ttest_1samp(sample['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

## Conclusion:  

Due to a t-statistic of -0.957, and a p-value of 0.364, we fail to reject the null hypothesis that the sample mean for highway-mpg is equal to 35mpg.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [24]:
sample2 = df.sample(100, random_state=30 )

sample2

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
..         ...               ...            ...  ...      ...         ...    ...
189          3               NaN     volkswagen  ...       24          29  11595
192          0               NaN     volkswagen  ...       33          38  13845
171          2               134         toyota  ...       24          30  11549
203         -1                95          volvo  ...       26          27  22470
47           0               145         jaguar  ...       15          19  32250

[100 rows x 26 columns]

In [25]:
stats.ttest_1samp(sample2['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

## Conclusion:  

Due to a t-statistic of -6.519, and a p-value of 0.00000000300, we reject the null hypothesis that the sample mean for highway-mpg is equal to 35mpg, and suggest the alternative hypothesis that the sample mean is not equal to what the salesman says (35 mpg).

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Because with larger datasets, we are able to be more certain of our results/significance.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [26]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-09-08 19:01:33--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-09-08 19:01:33 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [27]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [28]:
df = df.replace({'y': 1, 'n': 0, "?": np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [29]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [30]:
rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']

In [31]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [43]:
rep.tail()

party  handicapped-infants  ...  duty-free  south-africa
427  republican                  0.0  ...        0.0           1.0
430  republican                  0.0  ...        0.0           1.0
432  republican                  0.0  ...        0.0           1.0
433  republican                  0.0  ...        0.0           1.0
434  republican                  0.0  ...        NaN           0.0

[5 rows x 17 columns]

In [32]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [45]:
dem.tail()

party  handicapped-infants  ...  duty-free  south-africa
425  democrat                  0.0  ...        1.0           NaN
426  democrat                  1.0  ...        1.0           1.0
428  democrat                  NaN  ...        1.0           1.0
429  democrat                  1.0  ...        1.0           1.0
431  democrat                  0.0  ...        0.0           1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

##Null Hypothesis:
Republican support for the duty-free bill is equal to democrat levels of support for duty-free bill.

##Alternative Hypothesis:
Republican support for the duty-free bill is not equal to democrat levels of support for the duty-free bill.

##Confidence Level: 99%

In [34]:
stats.ttest_ind(dem['duty-free'], rep['duty-free'], nan_policy='omit')

Ttest_indResult(statistic=12.853146132542978, pvalue=5.997697174347365e-32)

In [37]:
dem['duty-free'].mean()

0.6374501992031872

In [38]:
rep['duty-free'].mean()

0.08974358974358974

##Conclusion:

Based on a t-statistic of 12.853, and a p-value of 5.998e-32, I reject the null hypothesis that republican levels of support are equal to democrat levels of support for the duty-free bill, and suggest the alternative hypothesis that the levels of support from each party are not equal.  

Based on the differences of the sample means and the sign of the t-statistic, it can be shown that democrat levels of support for the duty-free bill were significantly higher than the republican support for the same bill. 

##Null Hypothesis:
Republican support for the el-salvador-aid bill is equal to democrat levels of support for el-salvador-aid bill.

##Alternative Hypothesis:
Republican support for the el-salvador-aid bill is not equal to democrat levels of support for the el-salvador-aid bill.

##Confidence Level: 99%

In [47]:
stats.ttest_ind(dem['el-salvador-aid'], rep['el-salvador-aid'], nan_policy='omit')

Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

In [49]:
dem['el-salvador-aid'].mean()

0.21568627450980393

In [48]:
rep['el-salvador-aid'].mean()

0.9515151515151515

##Conclusion:

Based on a t-statistic of -21.137, and a p-value of 5.601e-68, I reject the null hypothesis that republican levels of support are equal to democrat levels of support for the el-salvador-aid bill, and suggest the alternative hypothesis that the levels of support from each party are not equal.  

Based on the differences of the sample means and the sign of the t-statistic, it can be shown that republican levels of support for the el-salvador-aid bill were significantly higher than the democrat levels of support for the same bill. 

##Null Hypothesis:
Republican support for the water-project bill is equal to democrat levels of support for water-project bill.

##Alternative Hypothesis:
Republican support for the water-project bill is not equal to democrat levels of support for the water-project bill.

##Confidence Level: 90%

In [50]:
stats.ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

In [51]:
dem['water-project'].mean()

0.502092050209205

In [52]:
rep['water-project'].mean()

0.5067567567567568

##Conclusion:

Based on a t-statistic of -0.089, and a p-value of 0.929, I fail to reject the null hypothesis that republican levels of support are equal to democrat levels of support for the water-project bill.

Based on the differences of the sample means and the sign of the t-statistic, it can be shown that republican levels of support for the water-project bill were relatively equal to the democrat levels of support for the same bill.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE